media/src/models.js

```diff
@@ -0,0 +1,159 @@
+                 lvlUp=true, viewGraph=true, rangeT=[0, 2], tangents=false,
+                 color='#000000', integrate=false, n=1, sample=0) {
+        this.f = f;
+        this.rangeX = rangeX;
+        this.rangeY = rangeY;
+        this.meshX = meshX;
+        this.meshY = meshY;
+        this.res = res;
+        this.lvlUp = lvlUp;
+        this.viewGraph = viewGraph;
+        this.rangeT = rangeT;
+        this.tangents = tangents;
+        this.color = color;
+        this.integrate = integrate;
+        this.n = n;
+        this.sample = sample;
+    }
+}
+
+class LevelSurface {
+    constructor (g='x^2 + y^2 - z^2', k=1, rangeX=[-2,2], rangeY=[-2, 2],
+                 rangeZ=[-2, 2], tangents=false) {
+        this.g = g;
+        this.k = k;
+        this.rangeX = rangeX;
+        this.rangeY = rangeY;
+        this.rangeZ = rangeZ;
+        this.tangents = tangents;
+    }
+}
+
+class Message {
+    constructor (message, sender) {
+        this.message = message;
+        this.sentBy = sender;
+        this.timestamp = Date.now();
+    }
+
+    get(prop) {
+        if (prop === 'time') {
+            return this.timestamp;
+        } else if (prop === 'user') {
+            return this.sentBy;
+        }
+    }
+}
+
+class ObjectMessage extends Message {
+    constructor (message, objectList, sender) {
+        super(message, sender);
+        this.objects = Array.from(objectList);
+    }
+}
+
+class ParametricSurface {
+    constructor (x='cos(u)*(1 + sin(v)/3)', y='sin(u)*(1 + sin(v)/3)',
+                 z='cos(v)/3', rangeU=['0', '2*pi'], rangeV=['0', '2*pi'],
+                 meshU=10, meshV=10, res=30) {
+        this.x = x;
+        this.y = y;
+        this.z = z;
+        this.rangeU = rangeU;
+        this.rangeV = rangeV;
+        this.meshU = meshU;
+        this.meshV = meshV;
+        this.res = res;
+    }
+}
+
+/**
+ * A basic container for holding and sorting responses.
+ */
+class ResponseList {
+    constructor (list) {
+        if (!Array.isArray(list)) {
+            try {
+                throw new Error('ResponseList must hold an array');
+            } catch (err) {
+                console.error(err);
+            }
+        } else {
+            this.list = list;
+        }
+        this.desc = false;
+        this.sortType = 'time';
+    }
+
+    /**
+     * It can be assumed that the initial list is sorted chronologically in
+     * the order that responses were received.
+     * @param {*} type The sorting type as a string: time, user
+     * @param {*} desc The sort direction; either ascending or descending
+     * @returns Sorted array of TextResponse || GraphResponse
+     */
+    sortBy (type = 'time', desc = false) {
+        const validSort = ['time', 'user'];
+        if (type === this.sortType && desc === this.desc) {
+            return;
+        } if (validSort.includes(type)) {
+            this.list = this.list.sort(function(a, b) {
+                return a.get(type) < b.get(type);
```

Comment:
So, I had the syntax wrong in my quick example I gave you. The correct sort function would use `-`, per the example [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#examples) like:
```
function(a, b) {
    return a.get(type) - b.get(type);
}
```
In addition, I don't think the objects we're sorting have a `get` method. If you want to get an arbitrary attribute from a JS object it would be `a[type] - b[type]`. Without tests or somewhere in the code that's using this, these errors wouldn't otherwise be uncovered, so it's a bit awkward to include code that's not in use yet. To move forward here I think we could make a `media/test/models.test.js` file with some tests like this:


```
 import assert from 'assert';
 import {
     ResponseList
 } from '../src/models.js';

 describe('ResponseList', function() {
     it('handles an empty list properly',
        function() {
            const sorted = new ResponseList([]).sortBy();
            assert.equal(sorted, []);
        });

     it('should sort an array of responses',
        function() {
            const sorted = new ResponseList([
                {
                    time: 124
                }, {
                    time: 930590235
                }
            ]).sortBy();

            // etc..                                                                                                                                                 
        });
 });
```

That said, if all this class is doing is handling some sorting, I'm not sure if it's necessary.
It's pretty straightforward to just sort an array of objects by arbitrary attributes in javascript.
Or maybe we just need to tie this code to some actual UI for it to really make sense.

Reply:
I'm actually in the process of writing tests, I'll add your tests to the list.

I figured a container class would store an `id` to connect the ResponseList to the Question or Task given by the professor, then we can store the ResponseLists in a list. It could even have a timestamp property so that that list of ResponseLists could be sorted and filtered chronologically. I also think that the container class allows to provide different sort options fairly easily without external code, just plug it in and feed it the parameters. 

I did implement a get function, but that might just be a hold over from Java. If all the properties are public then yeah, I can drop the function.